File: 01_conjunctive_pose_cells_network/yaw_height_hdc_network/create_yaw_height_hdc_weights.py
import numpy as np


def create_yaw_height_hdc_weights(yawDim, heightDim, yawVar, heightVar):
    """
    Creates a 2D normalized distribution of size (yawDim x heightDim) with
    variances yawVar and heightVar.

    Parameters:
        yawDim (int): The dimension for the yaw axis.
        heightDim (int): The dimension for the height axis.
        yawVar (float): The variance for the yaw axis.
        heightVar (float): The variance for the height axis.

    Returns:
        numpy.ndarray: The normalized 2D weight distribution.
    """
    yawDimCentre = (yawDim // 2) + 1
    heightDimCentre = (heightDim // 2) + 1

    # Create an empty 2D weight matrix
    weight = np.zeros((yawDim, heightDim))

    # Calculate the 2D Gaussian distribution
    for h in range(heightDim):
        for y in range(yawDim):
            weight[y, h] = (
                (1 / (yawVar * np.sqrt(2 * np.pi)))
                * np.exp(-((y - yawDimCentre) ** 2) / (2 * yawVar**2))
                * (1 / (heightVar * np.sqrt(2 * np.pi)))
                * np.exp(-((h - heightDimCentre) ** 2) / (2 * heightVar**2))
            )

    # Normalize the weights to ensure the sum equals 1
    total = np.sum(weight)
    weight /= total

    return weight


# Explanation:

#     yawDimCentre and heightDimCentre: These are computed to find the center of the yaw and height dimensions, similar to the MATLAB floor operation.
#     Weight Matrix Calculation: The 2D weight matrix is created using nested loops, where each element is calculated based on a 2D Gaussian distribution using yawVar and heightVar.
#     Normalization: The total sum of the weight matrix is computed using np.sum(), and then each element is divided by the total to normalize the matrix.
#     Return Value: The function returns the normalized 2D weight matrix.

# Example Usage:

# yawDim = 10
# heightDim = 10
# yawVar = 1.0
# heightVar = 1.0

# weights = create_yaw_height_hdc_weights(yawDim, heightDim, yawVar, heightVar)
# print(weights)

# This Python function should behave similarly to the MATLAB version and will generate a 2D normalized Gaussian distribution based on the specified dimensions and variances.

File: 01_conjunctive_pose_cells_network/yaw_height_hdc_network/create_yaw_height_hdc_weights.js
export function createYawHeightHdcWeights(yawDim, heightDim, yawVar, heightVar) {
    // Calculate the centers for yaw and height dimensions
    const yawDimCentre = Math.floor(yawDim / 2) + 1
    const heightDimCentre = Math.floor(heightDim / 2) + 1

    // Create an empty 2D array to store weights
    let weight = Array.from({ length: yawDim }, () => new Array(heightDim).fill(0))

    // Loop over all yaw and height values
    for (let h = 0; h < heightDim; h++) {
        for (let y = 0; y < yawDim; y++) {
            // Calculate the weight using the formula
            weight[y][h] = (1 / (yawVar * Math.sqrt(2 * Math.PI))) * Math.exp(-Math.pow(y - yawDimCentre, 2) / (2 * Math.pow(yawVar, 2))) * (1 / (heightVar * Math.sqrt(2 * Math.PI))) * Math.exp(-Math.pow(h - heightDimCentre, 2) / (2 * Math.pow(heightVar, 2)))
        }
    }

    // Normalize the weight array
    let total = weight.flat(Infinity).reduce((acc, val) => acc + val, 0) // Flatten the 2D array and sum all values
    for (let i = 0; i < yawDim; i++) {
        for (let j = 0; j < heightDim; j++) {
            weight[i][j] = weight[i][j] / total
        }
    }

    return weight
}
